File: fastapi-app/app/routers/widget.py
import re

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from sqlmodel import Session

from ..database import get_session
from ..models import Ticket
from ..rate_limit import rate_limit
from ..schemas import WidgetTicketCreate
from ..serializers import serialize_ticket
from ..services.webhook import send_ticket_created_webhook

router = APIRouter(prefix="/api/v2/widget", dependencies=[Depends(rate_limit(window_seconds=60, max_requests=5))])

EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
MAX_LEN = 4000


# Unauthenticated, rate-limited endpoint the public chat widget uses to file
# a ticket on a visitor's behalf. Deliberately narrower than the agent API
# in routers/tickets.py — it can only create a "chat" ticket, nothing else.
@router.post("/tickets.json", status_code=201)
def create_widget_ticket(
    body: WidgetTicketCreate,
    background_tasks: BackgroundTasks,
    session: Session = Depends(get_session),
):
    name = body.requester_name.strip()
    email = body.requester_email.strip()
    description = body.description.strip()

    if not name or not email or not description:
        raise HTTPException(
            status_code=400,
            detail={"title": "InvalidRecord", "message": "requester_name, requester_email, and description are required"},
        )
    if not EMAIL_RE.match(email):
        raise HTTPException(
            status_code=400,
            detail={"title": "InvalidRecord", "message": "requester_email must be a valid email address"},
        )
    if len(description) > MAX_LEN or len(name) > 200:
        raise HTTPException(status_code=400, detail={"title": "InvalidRecord", "message": "input too long"})

    subject = f"{description[:77]}..." if len(description) > 80 else description

    ticket = Ticket(
        subject=subject,
        description=description,
        channel="chat",
        requester_name=name,
        requester_email=email,
        status="new",
        tags="chat-widget",
        flag_type="[]",
    )
    session.add(ticket)
    session.commit()
    session.refresh(ticket)

    serialized = serialize_ticket(ticket)
    background_tasks.add_task(send_ticket_created_webhook, serialized)

    return {"ticket": serialized}

File: server/routes/widget.js
const express = require("express");
const prisma = require("../db");
const { sendTicketCreatedWebhook } = require("../services/webhook");
const { serializeTicket } = require("../ticketSerializer");

const router = express.Router();

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const MAX_LEN = 4000;

function badRequest(res, message) {
  return res.status(400).json({ error: { title: "InvalidRecord", message } });
}

// POST /api/v2/widget/tickets.json
// Unauthenticated, rate-limited endpoint the public chat widget uses to file
// a ticket on a visitor's behalf. Deliberately narrower than the agent API
// in routes/tickets.js — it can only create a "chat" ticket, nothing else.
router.post("/tickets.json", async (req, res, next) => {
  try {
    const body = req.body || {};
    const name = (body.requester_name || "").trim();
    const email = (body.requester_email || "").trim();
    const description = (body.description || "").trim();

    if (!name || !email || !description) {
      return badRequest(res, "requester_name, requester_email, and description are required");
    }
    if (!EMAIL_RE.test(email)) {
      return badRequest(res, "requester_email must be a valid email address");
    }
    if (description.length > MAX_LEN || name.length > 200) {
      return badRequest(res, "input too long");
    }

    const subject = description.length > 80 ? `${description.slice(0, 77)}...` : description;

    const ticket = await prisma.ticket.create({
      data: {
        subject,
        description,
        channel: "chat",
        requesterName: name,
        requesterEmail: email,
        status: "new",
        tags: "chat-widget",
        flagType: "[]",
      },
    });

    sendTicketCreatedWebhook(ticket).catch(() => {});

    res.status(201).json({ ticket: serializeTicket(ticket) });
  } catch (err) {
    next(err);
  }
});

module.exports = router;
